### PR TITLE
Refactor/control methods

### PR DESCRIPTION
**Pull request by Ratoone**
Refactored the control methods to inherit from a base controller. Classes that depend on the controllers still reference PID control - will be changed in the future, when integrating other controls.

As per @mrlukasbos suggestion, the branch has been cleared of old commits, resulting in a new branch (while deleting the old one). Also, the name of the new branch respects the branching name convention.

As per @HaicoDorenbos observation, javadoc has been added to the new classes

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -197,8 +200,7 @@ set(CONTROL_SOURCES
         ${PROJECT_SOURCE_DIR}/src/control/ballHandling/DribbleBackwards.cpp
         ${PROJECT_SOURCE_DIR}/src/control/ballHandling/DribbleForwards.cpp
         ${PROJECT_SOURCE_DIR}/src/control/ballHandling/RotateAroundBall.cpp
-        ${PROJECT_SOURCE_DIR}/src/control/ballHandling/RotateWithBall.cpp
-        )
+        ${PROJECT_SOURCE_DIR}/src/control/ballHandling/RotateWithBall.cpp)
```
Comment (rolfvdhulst) [inline CMakeLists.txt]: Not huge, but please put keep the ) on the next line. This prevents merge conflicts when multiple people are working on different branches.

File: include/roboteam_ai/control/PosController.h [hunk 2]
```diff
@@ -7,7 +7,8 @@
 
 #include "RobotCommand.h"
 #include <utilities/Constants.h>
-#include <control/pid.h>
+#include <include/roboteam_ai/control/controllers/PidController.h>
```
Comment (rolfvdhulst) [inline include/roboteam_ai/control/PosController.h]: You should never have to write include/roboteam_ai in your include's, you can remove it here (not that important but just convention)

File: src/control/PosController.cpp [hunk 3]
```diff
@@ -80,8 +71,8 @@ void PosController::setAutoListenToInterface(bool listenToInterface) {
 
 void PosController::updatePid(pidVals pid) {
     if (lastPid != pid) {
-        xpid.setPID(pid);
-        ypid.setPID(pid);
+        this->pid = PidTwoAxesController(std::tuple_cat(pid,std::make_tuple(0)));
```
Comment (rolfvdhulst) [inline src/control/PosController.cpp]: This probably needs a comment that explains what the 0 and std::tuple_cat are doing. (Something short like //we always set F to 0)

File: include/roboteam_ai/control/controllers/PidTwoAxesController.h [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+#define RTT_PIDTWOAXESCONTROLLER_H
+
+#include <include/roboteam_ai/control/ControlUtils.h>
+#include "BaseController.h"
+#include "PidController.h"
+
+class PidTwoAxesController : BaseController<Vector2,Vector2>{
+public:
+    PidController pidXAxis = PidController(0, 0, 0, 0);
+    PidController pidYAxis = PidController(0, 0, 0, 0);
```
Comment (rolfvdhulst) [inline include/roboteam_ai/control/controllers/PidTwoAxesController.h]: These probably do not need to be public.

File: include/roboteam_ai/control/controllers/PidTwoAxesController.h [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+    PidTwoAxesController(std::tuple<double, double, double, double> pidXParams, std::tuple<double, double, double, double> pidYParams);
+
+    explicit PidTwoAxesController(std::tuple<double, double, double, double> pidParams);
+
+    Vector2 computeOutput(Vector2 setPointInput, Vector2 actualInput) override;
+
+    PidTwoAxesController() = default;
```
Comment (rolfvdhulst) [inline include/roboteam_ai/control/controllers/PidTwoAxesController.h]: Another style/opinion thing but for most files we put the constructors/destructors as one of the first functions

File: include/roboteam_ai/control/PosController.h [hunk 6]
```diff
@@ -32,14 +33,11 @@ class PosController {
         bool customCanMoveInDefenseArea = true;
 
         // PID functions
-        PID xpid = PID(0.0, 0.0, 0.0);
-        PID ypid = PID(0.0, 0.0, 0.0);
+        PidTwoAxesController pid = PidTwoAxesController();
```
Comment (HaicoDorenbos) [inline include/roboteam_ai/control/PosController.h]: What is the relation between the PosController and PidTwoAxesController? It is unclear for me, so this variable needs a comment which explains this relation.

File: include/roboteam_ai/control/controllers/BaseController.h [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+#ifndef RTT_BASECONTROLLER_H
+#define RTT_BASECONTROLLER_H
+
+template <typename Input, typename Output>
+/**
+ * The base abstract class for the controllers. Any control method should inherit from this
+ * @tparam Input
+ * @tparam Output
+ */
```
Comment (HaicoDorenbos) [inline include/roboteam_ai/control/controllers/BaseController.h]: @author and @since are missing here

File: include/roboteam_ai/control/controllers/BaseController.h [hunk 8]
```diff
@@ -0,0 +1,25 @@
+//
+// Created by ratoone on 25-09-19.
+//
+
+#ifndef RTT_BASECONTROLLER_H
+#define RTT_BASECONTROLLER_H
+
+template <typename Input, typename Output>
+/**
+ * The base abstract class for the controllers. Any control method should inherit from this
```
Comment (HaicoDorenbos) [inline include/roboteam_ai/control/controllers/BaseController.h]: What is a controller and control method? This is still unclear for me when I read the documentation.

File: include/roboteam_ai/control/controllers/BaseController.h [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+ * @tparam Input
+ * @tparam Output
```
Comment (HaicoDorenbos) [inline include/roboteam_ai/control/controllers/BaseController.h]: The @param tag should not be used here, because the class does not use parameters only the function uses parameter.

File: include/roboteam_ai/control/controllers/PidTwoAxesController.h [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+//
+
+#ifndef RTT_PIDTWOAXESCONTROLLER_H
+#define RTT_PIDTWOAXESCONTROLLER_H
+
+#include <include/roboteam_ai/control/ControlUtils.h>
+#include "BaseController.h"
+#include "PidController.h"
+
+class PidTwoAxesController : BaseController<Vector2,Vector2>{
```
Comment (HaicoDorenbos) [inline include/roboteam_ai/control/controllers/PidTwoAxesController.h]: The PidTwoAxesController class and the public functions do not have any documentations. Also it is not clear for me what the Pid means. Does it mean "Pelvic inflammatory disease"? This should also be clear from the documentation or you should use the full name instead of the abbreviation Pid.

File: include/roboteam_ai/control/controllers/PidTwoAxesController.h [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+public:
+    PidController pidXAxis = PidController(0, 0, 0, 0);
```
Comment (HaicoDorenbos) [inline include/roboteam_ai/control/controllers/PidTwoAxesController.h]: Also not clear for me what these variables mean. But if you document the class then this might be better understandable.

File: src/control/PosController.cpp (hunk 3 above)
Comment (mrlukasbos) [reply src/control/PosController.cpp]: You can maybe also check for the existence of a fourth variable in the tuple in the constructor and put it to 0 there.

File: include/roboteam_ai/control/controllers/PidTwoAxesController.h (hunk 5 above)
Comment (HaicoDorenbos) [reply include/roboteam_ai/control/controllers/PidTwoAxesController.h]: I agree, however it is not a major issue.

Comment (HaicoDorenbos) [inline include/roboteam_ai/control/controllers/PidTwoAxesController.h]: These functions are public, so they should have documentation added as well.

File: src/control/controllers/PidTwoAxesController.cpp [hunk 12]
```diff
@@ -0,0 +1,36 @@
+//
```
Comment (HaicoDorenbos) [inline src/control/controllers/PidTwoAxesController.cpp]: Documentation should be added in the header file not in the source file.

File: src/control/controllers/PidTwoAxesController.cpp [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+// Created by ratoone on 25-09-19.
+//
+
+#include <control/controllers/PidTwoAxesController.h>
+
+/**
+ * Create a PID controller for two axes, with different parameters
+ * @param pidXParams a tuple with 4 parameters for X dir PID - proportional, integrative, derivative and feedforward gain
+ * @param pidYParams a tuple with 4 parameters for Y dir PID - proportional, integrative, derivative and feedforward gain
+ */
```
Comment (HaicoDorenbos) [inline src/control/controllers/PidTwoAxesController.cpp]: Documentation should be added in the header file not in the source file

File: src/control/controllers/PidTwoAxesController.cpp [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+PidTwoAxesController::PidTwoAxesController(std::tuple<double, double, double, double> pidXParams, std::tuple<double, double, double, double> pidYParams){
+    pidXAxis = PidController(pidXParams);
+    pidYAxis = PidController(pidYParams);
+}
+
+/**
+ * Computes the PID output for the two axes
+ * @param setPointInput the setpoint expressed as a Vector2
```
Comment (HaicoDorenbos) [inline src/control/controllers/PidTwoAxesController.cpp]: This param documentation does not give me any more information. I indeed know that it is a setpoint, because the variable is called a setPoint and I know that it is a Vector2, because the type of this variable is Vector2

File: src/control/controllers/PidTwoAxesController.cpp [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+ * @param pidXParams a tuple with 4 parameters for X dir PID - proportional, integrative, derivative and feedforward gain
+ * @param pidYParams a tuple with 4 parameters for Y dir PID - proportional, integrative, derivative and feedforward gain
+ */
+PidTwoAxesController::PidTwoAxesController(std::tuple<double, double, double, double> pidXParams, std::tuple<double, double, double, double> pidYParams){
+    pidXAxis = PidController(pidXParams);
+    pidYAxis = PidController(pidYParams);
+}
+
+/**
+ * Computes the PID output for the two axes
```
Comment (HaicoDorenbos) [inline src/control/controllers/PidTwoAxesController.cpp]: What is PID?

File: src/control/controllers/PidTwoAxesController.cpp [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+ * @param setPointInput the setpoint expressed as a Vector2
+ * @param actualInput the real value expressed as a Vector2
```
Comment (HaicoDorenbos) [inline src/control/controllers/PidTwoAxesController.cpp]: Same as above, this param documentation does not make anything more clear.

File: src/control/shotControllers/ShotController.cpp [hunk 17]
```diff
@@ -164,7 +164,7 @@ RobotCommand ShotController::moveStraightToBall(world::Robot robot,
     //auto newPid = interface::Output::getNumTreePid();
     auto newPidValues = interface::Output::getShotControllerPID();
     updatePid(newPidValues);
-    double pidOutput = pid.getOutput(err.length()*sign, 0);
+    double pidOutput = pid.computeOutput(err.length() * sign, 0);
```
Comment (HaicoDorenbos) [inline src/control/shotControllers/ShotController.cpp]: Indeed there should be a space between the multiplication operator and after the multiplication operator :+1:

File: include/roboteam_ai/control/ballHandling/BallHandlePosControl.h [hunk 18]
```diff
@@ -50,16 +49,14 @@ class BallHandlePosControl : public NumTreePosControl {
         int ticksNotMoving = 0;
 
         pidfVals pidfGoToBall = std::make_tuple(0.0, 0.0, 0.0, 1.5);
-        PID xGoToBallPID = PID(pidfGoToBall);
-        PID yGoToBallPID = PID(pidfGoToBall);
+        PidTwoAxesController goToBallPid = PidTwoAxesController(pidfGoToBall, pidfGoToBall);
```
Comment (Ratoone) [inline include/roboteam_ai/control/ballHandling/BallHandlePosControl.h]: @mrlukasbos The idea was to use dependency injection for the actual controller. So I have an interface (BaseController) which will be used instead of the implementation. For instance, we have here the PidTwoAxesController. Instead of it, maybe I want to use some other controller. Then I will just declare a BaseController<Vector2,Vector2>, and through some setter / in the constructor I can set which type of controller we want

Comment (mrlukasbos) [reply include/roboteam_ai/control/ballHandling/BallHandlePosControl.h]: I understand. Why not set the controller you want immediately? Now you are forced to use generics making your code significantly harder.

Btw, this is very much opinion based. Its your project now and if you want it like this then don't let me stop you from doing so :-)

File: include/roboteam_ai/control/controllers/PidTwoAxesController.h (hunk 10 above)
Comment (rolfvdhulst) [reply include/roboteam_ai/control/controllers/PidTwoAxesController.h]: I disagree with this sentiment to be honest. It is googleable and anyone working on this code is very likely to know what it means (and actually does need to know in order to fully understand/edit the code), so this should be common knowledge.

File: src/control/controllers/PidTwoAxesController.cpp (hunk 15 above)
Comment (rolfvdhulst) [reply src/control/controllers/PidTwoAxesController.cpp]: https://en.m.wikipedia.org/wiki/PID_controller

File: src/control/PosController.cpp (hunk 3 above)
Comment (rolfvdhulst) [reply src/control/PosController.cpp]: That would be even prettier.

File: include/roboteam_ai/control/controllers/PidTwoAxesController.h (hunk 10 above)
Comment (HaicoDorenbos) [reply include/roboteam_ai/control/controllers/PidTwoAxesController.h]: I think it is still better to add class documentation, because of the following reasons:
1. When I search for Pid Two Axes Controller I cannot find anything that directly explains how it works. 
2. Documentation of code is a major issue of the RoboTeam.
3. When you add documentation to code we can quickly using CTRL-Q check how a function works. 

If there is some online source which describes how something works then the URL to this source should be included in the documentation.

File: src/control/controllers/PidTwoAxesController.cpp (hunk 15 above)
Comment (HaicoDorenbos) [reply src/control/controllers/PidTwoAxesController.cpp]: If this URL clearly explains how it works then I propose to add this URL to the documentation.

File: include/roboteam_ai/control/controllers/BaseController.h (hunk 8 above)
Comment (RobotJesse) [reply include/roboteam_ai/control/controllers/BaseController.h]: I don't think we need this much detail here, everyone in the control team should know what a controller is

**Comment by Ratoone** [comment]
Updated with the requested changes

File: include/roboteam_ai/control/controllers/BaseController.h [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+#define RTT_BASECONTROLLER_H
+
+template <typename Input, typename Output>
+/**
+ * The base abstract class for the controllers (e.g. velocity control, position control). Any control
+ * method (e.g. PID control) should inherit from this.
+ * @tparam Input the type of the input of the controller (reference type)
+ * @tparam Output the type of the output of the controller (plant input)
+ * @author Paul Vacariu
+ * @SInCE 25.09.2019
```
Comment (HaicoDorenbos) [inline include/roboteam_ai/control/controllers/BaseController.h]: @since should be written in lowercase

**Comment by HaicoDorenbos** [comment]
Level of documentation seems to me to be quite satisfactory, but I will still do a full check...

File: include/roboteam_ai/control/controllers/PidController.h [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+//
+
+#ifndef ROBOTEAM_AI_PID_H
+#define ROBOTEAM_AI_PID_H
+
+#include <tuple>
+#include "BaseController.h"
+
+/**
+ * The basic implementation of a PIDF (proportional - integrative - derivative
```
Comment (HaicoDorenbos) [inline include/roboteam_ai/control/controllers/PidController.h]: Why do you put - between the words and why do you have a line break here?

File: include/roboteam_ai/control/controllers/PidController.h [hunk 21]
```diff
@@ -0,0 +1,144 @@
...
+ * - feedback) controller. The actual controller accepts a float reference
+ * and outputs a float controller action. For more info check https://en.wikipedia.org/wiki/PID_controller
```
Comment (HaicoDorenbos) [inline include/roboteam_ai/control/controllers/PidController.h]: It is now more clear for me, because I know this class is not about the Pelvic Inflammatory Disease. Reference URL to the Wikipedia page which explains more is quite useful.

File: include/roboteam_ai/control/controllers/PidController.h [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+     * @param d Derivative gain. Responds quickly to large changes in error. Small values prevents P and I terms from causing overshoot.
+     */
+    PidController(double p, double i, double d);
+    /** Create a new PID object.
+     * @param p Proportional gain. Large if large difference between setpoint and target.
+     * @param i Integral gain.	Becomes large if setpoint cannot reach target quickly.
+     * @param d Derivative gain. Responds quickly to large changes in error. Small values prevents P and I terms from causing overshoot.
+     * @param f Feedforward gain. The value that will be multiplied with the input and added to the output
+     */
+    PidController(double p, double i, double d, double f);
```
Comment (HaicoDorenbos) [inline include/roboteam_ai/control/controllers/PidController.h]: Good job :+1: we should indeed know what the parameters are used for in the constructor

File: include/roboteam_ai/control/controllers/PidController.h [hunk 23]
```diff
@@ -0,0 +1,144 @@
...
+    /** Create a new PID object
+     * @oaram pid tuple with the three parameters of the PID
```
Comment (HaicoDorenbos) [inline include/roboteam_ai/control/controllers/PidController.h]: What is an oaram?

File: include/roboteam_ai/control/controllers/PidController.h [hunk 24]
```diff
@@ -0,0 +1,144 @@
...
+     */
+    explicit PidController(std::tuple<double, double, double> pid);
+    /** Create a new PID object
+     * @oaram pid tuple with the three parameters of the PID, plus feedforward gain
```
Comment (HaicoDorenbos) [inline include/roboteam_ai/control/controllers/PidController.h]: Like before, this is misspelled

File: include/roboteam_ai/control/controllers/PidController.h [hunk 25]
```diff
@@ -0,0 +1,144 @@
...
+     * @param max Maximum value of range
+     * @return
+     */
+    bool bounded(double, double, double);
+    /**
+     * To operate correctly, all PID parameters require the same sign,
+     * with that sign depending on the {@literal}reversed value
+     */
+    void checkSigns();
+    double P;
```
Comment (HaicoDorenbos) [inline include/roboteam_ai/control/controllers/PidController.h]: It is better to use the full names instead, like proportionalGain to make these variables more meaningfull

File: include/roboteam_ai/control/controllers/PidController.h [hunk 26]
```diff
@@ -0,0 +1,144 @@
...
+    double maxIOutput;
+    double maxError;
+    double errorSum;
+
+    double maxOutput;
+    double minOutput;
+
+    double setpoint;
+
+    double lastActual;
```
Comment (HaicoDorenbos) [inline include/roboteam_ai/control/controllers/PidController.h]: Some of these variables can be const

File: include/roboteam_ai/control/controllers/PidController.h [hunk 27]
```diff
@@ -0,0 +1,144 @@
...
+    /** Create a new PID object
+     * @oaram pid tuple with the three parameters of the PID
+     */
+    explicit PidController(std::tuple<double, double, double> pid);
+    /** Create a new PID object
+     * @oaram pid tuple with the three parameters of the PID, plus feedforward gain
+     */
+    explicit PidController(std::tuple<double, double, double, double> pidf);
+
+    void setPID(double, double, double);
```
Comment (HaicoDorenbos) [inline include/roboteam_ai/control/controllers/PidController.h]: Huh?! Where are the parameter names?

File: include/roboteam_ai/control/controllers/PidController.h [hunk 28]
```diff
@@ -0,0 +1,144 @@
...
+     * @param maximum. Units are the same as the expected output value
+     */
+    void setMaxIOutput(double maximum);
+    /**Specify a maximum output. If a single parameter is specified, the minimum is
+     * set to (-maximum).
+     * @param output
+     */
+    void setOutputLimits(double output);
+    /**
+     * Specify a maximum output.
```
Comment (HaicoDorenbos) [inline include/roboteam_ai/control/controllers/PidController.h]: I guess you also specify minimum output, so this should be included in the summary as well

File: include/roboteam_ai/control/controllers/PidController.h [hunk 29]
```diff
@@ -0,0 +1,144 @@
...
+     * @param minimum possible output value
+     * @param maximum possible output value
+     */
+    void setOutputLimits(double minimum, double maximum);
+    /** Set the operating direction of the PID controller
+     * @param reversed Set true to reverse PID output
+     */
+    void setDirection(bool reversed);
```
Comment (HaicoDorenbos) [inline include/roboteam_ai/control/controllers/PidController.h]: So if you do twice setDirection(true) you will move in the same direction as you did before?

File: include/roboteam_ai/control/controllers/PidController.h [hunk 30]
```diff
@@ -0,0 +1,144 @@
...
+    /**
+     * Resets the controller. this->erases the I term buildup, and removes D gain on the next loop.
+     */
+    void reset();
+    /**Set the maximum rate the output can increase per cycle.
+     * @param rate
```
Comment (HaicoDorenbos) [inline include/roboteam_ai/control/controllers/PidController.h]: Missing explanation

File: include/roboteam_ai/control/controllers/PidTwoAxesController.h [hunk 31]
```diff
@@ -0,0 +1,96 @@
...
+     * @param pidYParams a tuple with 4 parameters for Y dir PID - proportional, integrative, derivative and feedforward gain
+     */
+    PidTwoAxesController(std::tuple<double, double, double, double> pidXParams, std::tuple<double, double, double, double> pidYParams);
+
+    /**
+     * Create a PID controller for two axes, with different parameters and zero feedforward
+     * @param pidXParams a tuple with 3 parameters for X dir PID - proportional, integrative, derivative
+     * @param pidYParams a tuple with 3 parameters for Y dir PID - proportional, integrative, derivative
+     */
+    PidTwoAxesController(std::tuple<double, double, double> pidXParams, std::tuple<double, double, double> pidYParams);
```
Comment (HaicoDorenbos) [inline include/roboteam_ai/control/controllers/PidTwoAxesController.h]: It is also possible in C++ to use default arguments and then you can remove this overloaded constructor and use the previous constructor for this case

File: src/control/controllers/PidController.cpp [hunk 32]
```diff
@@ -0,0 +1,225 @@
...
+PidController::PidController(std::tuple<double, double, double> pid)
+        : PidController(std::get<0>(pid), std::get<1>(pid), std::get<2>(pid)) {}
+
+PidController::PidController(std::tuple<double, double, double, double> pidf)
+        : PidController(std::get<0>(pidf), std::get<1>(pidf), std::get<2>(pidf), std::get<3>(pidf)) {}
+
+
+//**********************************
+//Configuration functions
+//**********************************
```
Comment (HaicoDorenbos) [inline src/control/controllers/PidController.cpp]: In the original pid.cpp file we had more documentation here, why did you remove it?

File: src/control/controllers/PidController.cpp [hunk 33]
```diff
@@ -0,0 +1,225 @@
...
+    setpoint=0;
+    lastActual=0;
+    firstRun=true;
+    reversed=false;
+    outputRampRate=0;
+    lastOutput=0;
+    outputFilter=0;
+    setpointRange=0;
+
+    P=p; I=i; D=d; F=f;
```
Comment (HaicoDorenbos) [inline src/control/controllers/PidController.cpp]: According to C++ policies you should never have multiple code lines at the same line, after a ; there should be a new line.

File: src/control/controllers/PidTwoAxesController.cpp [hunk 34]
```diff
@@ -0,0 +1,46 @@
...
+        : PidTwoAxesController(pidParams, pidParams){}
+
+Vector2 PidTwoAxesController::computeOutput(Vector2 setPointInput, Vector2 actualInput) {
+    Vector2 outputCommand;
+    outputCommand.x = pidXAxis.computeOutput(setPointInput.x, actualInput.x);
+    outputCommand.y = pidYAxis.computeOutput(setPointInput.y, actualInput.y);
+    return outputCommand;
+}
+
+void PidTwoAxesController::setOutputLimits(double minimumX, double maximumX, double minimumY, double maximumY) {
```
Comment (HaicoDorenbos) [inline src/control/controllers/PidTwoAxesController.cpp]: Nice Facade pattern

File: include/roboteam_ai/control/controllers/PidController.h (hunk 29 above)
Comment (Ratoone) [reply include/roboteam_ai/control/controllers/PidController.h]: Added an explanation to the summary taken from the code. Apparently the function is never used though
